designs/0003-ragged-containers.md

```diff
@@ -0,0 +1,575 @@
+then `x` is said to be *rectangular*.
+
+We write the unsized type of `x` the same way as for rectangular two-dimensional arrays,
+
+```
+x : real[ , ]
+```
+
+Declarations for ragged arrays are made up of the sizes of their
+elements.  Declaring `x` of type and size `real[{3, 4}]` declares `x`
+to be a two-dimensional array with `x[1]` having size three and `x[2]`
+size four.
+
+Providing an out-of-bounds index will raise the same exceptions (and
+thus cause the same behavior in interfaces) as other out-of-bounds
+errors.
+
+A ragged three-dimensional array may be defined similarly; each
+element in the sizing is the sizing of a ragged two-dimensional array.
+
+```
+real[{ {2, 3}, { 1, 2, 3 } }] y
+  = { {{a, b, c}, {d, e, f, g}},
+      {{h}, {i, j}, {k, l, m}} };
+```
+
+The size of `y` is the number of elements in its top-level
+declaration, here `size(y) == 2`.  The first element of `y` is a
+ragged two-dimensional array,
+
+```
+y[1] == {{a, b, c}, {d, e, f, g}}
+```
+
+and the second is
+
+```
+y[2] == {{h}, {i, j}, {k, l, m}}
+```
+
+The type of both `y[1]` and `y[2]` is `real[, ]`, the type of a
+real-valued two-dimensional array.  This is a requirement of ragged
+structures.
+
+This brings up an important principle of ragged structures:  each
+element in a ragged structure must be the same unsized type.  That is,
+it's not possible to have a ragged array whose first element is
+two dimensional and whose second element is three dimensional.
+
+### Ragged arrays of vectors
+
+In addition to arrays, the values of ragged structures may be vectors
+or matrices.  For example, a ragged array of vectors may be declared
+and defined as
+
+```
+vector[{3, 2}] v = { [a, b c]', [e, f]' };
+```
+
+The variable `v` is typed as an array of vectors, `v : vector[]`.  The
+value of `v[1]` is the size three vector `[a, b, c]'`, whereas the
+value of `v[2]` is the size two vector `[e, f]'`.  Both `v[1]` and
+`v[2]` are of type `vector`.  All 
+
+All size functions (`rows()`, `cols()`, etc.) work as expected.
+
+Row vectors are handled identically, e.g.,
+
+```
+row_vector[{3, 2}] w = { [a, b c], [e, f] };
+```
+
+### Ragged arrays of matrices
+
+
+Ragged arrays of matrices are more challenging to declare because
+their base elements are two dimensional.  For example,
+
+```
+matrix[2, 3] u = [[a, b, c], [d, e, f]];
+```
+
+declares `u` to be a two by three matrix.  In order to deal with
+ragged matrix declarations in general, the following declaration is
+identical to the previous, for a two by three matrix,
+
+
+```
+matrix[{2, 3}] u = [[a, b, c], [d, e, f]];
+```
+
+This is chosen to match the dimensions declaration, where the function
+`dims` returns a matrix's dimensions,
+
+```
+dims(u) == {2, 3}
+```
+
+A ragged array of matrices is declared using a sequence of sizes.  For
+example,
+
+```
+matrix[{ {2, 3}, {3, 2}, {1, 2} }] v
+ = { [[a, b, c], [d, e, f]],
+     [[g, h], [i, j], [k, l]],
+     [[m, n]] };
```

Comment:
```suggestion
array[3] matrix[{ {2, 3}, {3, 2}, {1, 2} }] v
 = { [[a, b, c], [d, e, f]],
     [[g, h], [i, j], [k, l]],
     [[m, n]] };
```

Reply:
Optional: 

Mention that 

`matrix[{2, 3}] u = [[a, b, c], [d, e, f]];` is a special case of `array[1] matrix[{2, 3}] u = [[a, b, c], [d, e, f]];`

which is why `matrix[{2,3}]` is the same as `matrix[2,3]`. At least to me, seeing it this way made it easier to understand why `matrix[2,3] = matrix[{2,3}]`.